presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedInPredicateToJoin.java

```diff
@@ -0,0 +1,425 @@
+    {
+        return new JoinNode(
+                idAllocator.getNextId(),
+                JoinNode.Type.LEFT,
+                probeSide,
+                buildSide,
+                ImmutableList.of(),
+                ImmutableList.<Symbol>builder()
+                        .addAll(probeSide.getOutputSymbols())
+                        .addAll(buildSide.getOutputSymbols())
+                        .build(),
+                Optional.of(joinExpression),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+    }
+
+    private static AggregationNode.Aggregation countWithFilter(Expression condition)
+    {
+        FunctionCall countCall = new FunctionCall(
+                QualifiedName.of("count"),
+                Optional.<Window>empty(),
+                Optional.of(condition),
+                false,
+                ImmutableList.<Expression>of() /* arguments */
+        );
+
+        return new AggregationNode.Aggregation(
+                countCall,
+                new Signature("count", FunctionKind.AGGREGATE, BIGINT.getTypeSignature()),
+                Optional.<Symbol>empty() /* mask */
+        );
+    }
+
+    private static Expression isGreaterThan(Symbol symbol, long value)
+    {
+        return new ComparisonExpression(
+                ComparisonExpressionType.GREATER_THAN,
+                symbol.toSymbolReference(),
+                bigint(value)
+        );
+    }
+
+    private static Expression not(Expression booleanExpression)
+    {
+        return new NotExpression(booleanExpression);
+    }
+
+    private static Expression isNotNull(Symbol symbol)
+    {
+        return new IsNotNullPredicate(symbol.toSymbolReference());
+    }
+
+    private static Expression bigint(long value)
+    {
+        return new Cast(new LongLiteral(String.valueOf(value)), BIGINT.toString());
+    }
+
+    private static Expression booleanConstant(@Nullable Boolean value)
+    {
+        if (value == null) {
+            return new Cast(new NullLiteral(), BOOLEAN.toString());
+        }
+        return new BooleanLiteral(value.toString());
+    }
+
+    /**
+     * TODO consult comon parts with {@link com.facebook.presto.sql.planner.optimizations.TransformCorrelatedScalarAggregationToJoin.Rewriter#decorrelateFilters}
+     */
+    private static class DecorrelatingVisitor
+            extends PlanVisitor<PlanNode, Optional<Decorrelated>>
+    {
+        private final Lookup lookup;
+        private final Set<Symbol> correlation;
+
+        public DecorrelatingVisitor(Lookup lookup, Iterable<Symbol> correlation)
+        {
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            this.correlation = ImmutableSet.copyOf(requireNonNull(correlation, "correlation is null"));
+        }
+
+        public Optional<Decorrelated> decorrelate(PlanNode reference)
```

Comment:
It has to be by nature how Rules work with Lookup, but the calling sites would need to cast which I'd prefer to avoid. Let's remove the check. I think other rules don't bother, they just `lookup` when navigating the plan.